File: Generate Parentheses - GFG/generate-parentheses.py
#User function Template for python3

class Solution:
    def AllParenthesis(self,n):
        #code here
        self.ans = []
        s = ""
        self.generate(s, n, n)
        return self.ans

    def generate(self, s, o, c):
        if o == 0 and c == 0:
            self.ans.append(s)
            return
        if o > 0:
            s += '('
            self.generate(s, o - 1, c)
            s = s[:-1]
        if c > 0:
            if o < c:
                s += ')'
                self.generate(s, o, c - 1)
                s = s[:-1]




#{ 
 # Driver Code Starts
#Initial Template for Python 3


        
if __name__=="__main__":
    t=int(input())
    for i in range(0,t):
        n=int(input())
        ob=Solution()
        result=ob.AllParenthesis(n)
        result.sort()
        for i in range(0,len(result)):
            print(result[i])
        

# } Driver Code Ends

File: Generate Parentheses - GFG/generate-parentheses.cpp
//{ Driver Code Starts
//Initial Template for C++

#include <bits/stdc++.h>
using namespace std;
vector<string> AllParenthesis(int n) ;


// } Driver Code Ends
//User function Template for C++

// N is the number of pairs of parentheses
// Return list of all combinations of balanced parantheses
class Solution
{
    vector<string>res;
    public:
    vector<string> AllParenthesis(int n) 
    {
        // Your code goes here 
    
        string s="";
        generate(s,n,n);
        return res;
    }
    void generate(string s,int o,int c){
        if (o==0 && c==0){
            res.push_back(s);
            return;
        }
        if (o>0){
            s.push_back('(');
            generate(s,o-1,c);
            s.pop_back();
        }
        if (c>0){
            if (o<c){
                s.push_back(')');
                generate(s,o,c-1);
                s.pop_back();
            }
        }
    }
};

//{ Driver Code Starts.


int main() 
{ 
	int t;
	cin>>t;
	while(t--)
	{
		int n;
		cin>>n;
		Solution ob;
		vector<string> result = ob.AllParenthesis(n); 
		sort(result.begin(),result.end());
		for (int i = 0; i < result.size(); ++i)
			cout<<result[i]<<"\n";
	}
	return 0; 
} 

// } Driver Code Ends